diff --git a/crates/conductor_lib/Cargo.toml b/crates/conductor_lib/Cargo.toml
index 449fbdd8f..514d0f3fa 100644
--- a/crates/conductor_lib/Cargo.toml
+++ b/crates/conductor_lib/Cargo.toml
@@ -52,9 +52,9 @@ nickel = "=0.11.0"
 url = { version = "=2.1.0", features = ["serde"] }
 
 [dev-dependencies]
-test_utils = { version = "=0.0.37-alpha3", path = "../../test_utils" }
+# test_utils = { version = "=0.0.37-alpha3", path = "../../test_utils" }
 tempfile = "=3.0.7"
-holochain_wasm_utils = { version = "=0.0.37-alpha3", path = "../wasm_utils" }
+# holochain_wasm_utils = { version = "=0.0.37-alpha3", path = "../wasm_utils" }
 structopt = "=0.2.15"
 pretty_assertions = "=0.6.1"
 ws = "=0.8.0"
diff --git a/crates/core/Cargo.toml b/crates/core/Cargo.toml
index 8ed726287..3710d527e 100644
--- a/crates/core/Cargo.toml
+++ b/crates/core/Cargo.toml
@@ -61,5 +61,5 @@ rand = "0.7.2"
 
 [dev-dependencies]
 wabt = "=0.7.4"
-test_utils = { version = "=0.0.37-alpha3", path = "../../test_utils" }
+# test_utils = { version = "=0.0.37-alpha3", path = "../../test_utils" }
 tempfile = "=3.0.7"
diff --git a/crates/core_types/Cargo.toml b/crates/core_types/Cargo.toml
index fa991369c..7b2bcb28a 100644
--- a/crates/core_types/Cargo.toml
+++ b/crates/core_types/Cargo.toml
@@ -47,5 +47,5 @@ log = "=0.4.8"
 holochain_logging = "=0.0.4"
 
 [dev-dependencies]
-test_utils = { version = "=0.0.37-alpha3", path = "../../test_utils"}
+# test_utils = { version = "=0.0.37-alpha3", path = "../../test_utils"}
 maplit = "=1.0.1"
diff --git a/crates/hdk/Cargo.toml b/crates/hdk/Cargo.toml
index 24285a7db..45e096566 100644
--- a/crates/hdk/Cargo.toml
+++ b/crates/hdk/Cargo.toml
@@ -24,9 +24,9 @@ env_logger = "=0.6.1"
 url = "=2.1.0"
 
 [dev-dependencies]
-test_utils = { version = "=0.0.37-alpha3", path = "../../test_utils" }
-holochain_conductor_lib = { version = "=0.0.37-alpha3", path = "../conductor_lib" }
-holochain_core = { version = "=0.0.37-alpha3", path = "../core" }
-holochain_core_types = { version = "=0.0.37-alpha3", path = "../core_types" }
+# test_utils = { version = "=0.0.37-alpha3", path = "../../test_utils" }
+# holochain_conductor_lib = { version = "=0.0.37-alpha3", path = "../conductor_lib" }
+# holochain_core = { version = "=0.0.37-alpha3", path = "../core" }
+# holochain_core_types = { version = "=0.0.37-alpha3", path = "../core_types" }
 tempfile = "=3.0.7"
 boolinator = "=2.4.0"
diff --git a/crates/wasm_utils/Cargo.toml b/crates/wasm_utils/Cargo.toml
index 64450ffe5..b742976b2 100644
--- a/crates/wasm_utils/Cargo.toml
+++ b/crates/wasm_utils/Cargo.toml
@@ -17,7 +17,7 @@ holochain_persistence_api = "=0.0.10"
 holochain_json_api = "=0.0.17"
 
 [dev-dependencies]
-test_utils = { version = "=0.0.37-alpha3", path = "../../test_utils" }
-holochain_core = { version = "=0.0.37-alpha3", path = "../core" }
-holochain_conductor_lib = { version = "=0.0.37-alpha3", path = "../conductor_lib" }
+# test_utils = { version = "=0.0.37-alpha3", path = "../../test_utils" }
+# holochain_core = { version = "=0.0.37-alpha3", path = "../core" }
+# holochain_conductor_lib = { version = "=0.0.37-alpha3", path = "../conductor_lib" }
 tempfile = "=3.0.7"
